README.md

# Display ClinVar variants in HiGlass

![ClinVar](https://aveit.s3.amazonaws.com/higlass/static/higlass-clinvar-screenshot.png)


**Note**: This is the source code for the ClinVar track only! You might want to check out the following repositories as well:

- HiGlass viewer: https://github.com/higlass/higlass
- HiGlass server: https://github.com/higlass/higlass-server
- HiGlass docker: https://github.com/higlass/higlass-docker

## Installation
 
```
npm install higlass-clinvar
```

## Data preparation

In order to bring a ClinVar vcf file (https://ftp.ncbi.nlm.nih.gov/pub/clinvar/vcf_GRCh38/) into the required BED format, the following script can be used (make sure to adjust the file names in the script). A chrom sizes file is required.
```
python /scripts/extract_clinvar_data.py
```

The resulting bed file can be aggregated and ingested into a HiGlass server in the usual way:
```
clodius aggregate bedfile --chromsizes-filename hg38.txt --delimiter $'\t' --importance-column 6 --max-per-tile 80 clinvar_20200824.bed 

python manage.py ingest_tileset --filename data/clinvar_20200824.beddb --filetype beddb --datatype bedlike --uid clinvar_20200824

```

## Usage

The live script can be found at:

- https://unpkg.com/higlass-clinvar/dist/higlass-clinvar.js

### Client

1. Load this track before the HiGlass core script. For example:

```
<script src="/higlass-clinvar.js"></script>
<script src="hglib.js"></script>

<script>
  ...
</script>
```

### Options
The following options are available:
```
{
  "server": "http://localhost:8001/api/v1",
  "tilesetUid": "awesome_orthologs",
  "uid": "awesome_orthologs_uid",
  "type": "horizontal-orthologs",
  "options": {
    "fontSize": 10,
    "fontFamily": "Arial",
    "labelTextColor": "#888888",
    "levelDistance": 20, // Detemines how far the different levels are visually apart
    "significanceColors": {
      "pathogenic": "#ff0000",
      "pathogenic_likely_pathogenic": "#ff3838",
      "likely_pathogenic": "#a80000",
      "uncertain_significance": "#808080",
      "likely_benign": "#009600",
      "benign_likely_benign": "#00c900",
      "benign": "#00ff00",
      "risk_factor": "#999999",
    },
  },
  "width": 768,
  "height": 200
}
```

### ECMAScript Modules (ESM)

We also build out ES modules for usage by applications who may need to import or use `higlass-clinvar` as a component.

Whenever there is a statement such as the following, assuming `higlass-clinvar` is in your node_modules folder:
```javascript
import { ClinvarTrack } from 'higlass-clinvar';
```

Then ClinvarTrack would automatically be imported from the `./es` directory (set via package.json's `"module"` value). 

## Support

For questions, please either open an issue or ask on the HiGlass Slack channel at http://bit.ly/higlass-slack

## Development

### Testing

To run the test suite:

```
npm run test-watch
```

### Installation

```bash
$ git clone https://github.com/higlass/higlass-clinvar.git
$ cd higlass-clinvar
$ npm install
```
If you have a local copy of higlass, you can then run this command in the higlass-clinvar directory:

```bash
npm link higlass
```

### Commands

 - **Developmental server**: `npm start`
 - **Production build**: `npm run build`



CHANGELOG.md

**v0.2.1**

- Smoother scrolling accross tiles

**v0.2.0**

- Use color-safe palette

**v0.1.0**

- Initial release


test/ClinvarTests.js
import { expect } from 'chai';
import register from 'higlass-register';

import FetchMockHelper from './utils/FetchMockHelper';

import { HiGlassComponent, getTrackObjectFromHGC } from 'higlass';

import {
  waitForDataLoaded,
  mountHGComponent,
  removeHGComponent,
} from './utils/test-helpers';

import viewConf from './view-configs/simple-track';

import ClinvarTrack from '../src/scripts/ClinvarTrack';

register({
  name: 'ClinvarTrack',
  track: ClinvarTrack,
  config: ClinvarTrack.config,
});

describe('Clinvar test', () => {
  const fetchMockHelper = new FetchMockHelper('', 'ClinvarTest');

  beforeAll(async () => {
    await fetchMockHelper.activateFetchMock();
  });

  describe('Test that the data is correct', () => {
    let hgc = null;
    let div = null;

    beforeAll((done) => {
      [div, hgc] = mountHGComponent(div, hgc, viewConf, done);
    });

    it('Test label data', (done) => {
      const trackObj = getTrackObjectFromHGC(
        hgc.instance(),
        viewConf.views[0].uid,
        viewConf.views[0].tracks.top[0].uid,
      );

      const sigLevels = trackObj.significanceLevels;
      expect(Object.keys(sigLevels).length).to.equal(9);

      done();
    });

    it('Test clinvar data', (done) => {
      // We need to wait for Ensemble data
      setTimeout(() => {
        const trackObj = getTrackObjectFromHGC(
          hgc.instance(),
          viewConf.views[0].uid,
          viewConf.views[0].tracks.top[0].uid,
        );

        const tile = trackObj.visibleAndFetchedTiles()[0];

        const rects = tile.rectsForMouseOver;

        const data = rects[0].data;

        expect(data.chrom).to.equal('chr1');
        expect(data.posRel).to.equal(2027636);
        expect(data.ref).to.equal('A');
        expect(data.alt).to.equal('C');
        expect(data.significance).to.equal('risk_factor');

        done();
      }, 2000);
    });

    afterAll(() => {
      removeHGComponent(div);
    });
  });

  afterAll(async () => {
    await fetchMockHelper.storeDataAndResetFetchMock();
  });
});


test/SVGExportTests.js
import { expect } from 'chai';
import register from 'higlass-register';

import FetchMockHelper from './utils/FetchMockHelper';

import { HiGlassComponent, getTrackObjectFromHGC } from 'higlass';

import {
  waitForDataLoaded,
  mountHGComponent,
  removeHGComponent,
} from './utils/test-helpers';

import viewConf from './view-configs/simple-track';

import ClinvarTrack from '../src/scripts/ClinvarTrack';

register({
  name: 'ClinvarTrack',
  track: ClinvarTrack,
  config: ClinvarTrack.config,
});

describe('SVG export', () => {
  const fetchMockHelper = new FetchMockHelper('', 'SVGExport');

  beforeAll(async () => {
    await fetchMockHelper.activateFetchMock();
  });

  describe('SVG export', () => {
    let hgc = null;
    let div = null;

    beforeAll((done) => {
      [div, hgc] = mountHGComponent(div, hgc, viewConf, done);
    });

    it('tests that the export works and the data is correct', (done) => {
      const trackObj = getTrackObjectFromHGC(
        hgc.instance(),
        viewConf.views[0].uid,
        viewConf.views[0].tracks.top[0].uid,
      );

      setTimeout(() => {
        hgc.instance().handleExportSVG();

        const tile = trackObj.visibleAndFetchedTiles()[0];

        // For some reason this passes locally, but fails on Travis
        // expect(tile.svgData[0].posX).to.equal(505.55494416122536);

        // expect(tile.svgData[0].posY).to.equal(96.5);
        // expect(tile.svgData[0].yZero).to.equal(56);
        // expect(tile.svgData[0].color).to.equal("#999999");

        // expect(tile.svgData[9].posX).to.equal(306.4364296378897);
        // expect(tile.svgData[9].posY).to.equal(76.5);
        // expect(tile.svgData[9].yZero).to.equal(56);
        // expect(tile.svgData[9].color).to.equal("#009600");

        done();
      }, 2000);
    });

    afterAll(() => {
      removeHGComponent(div);
    });
  });

  afterAll(async () => {
    await fetchMockHelper.storeDataAndResetFetchMock();
  });
});


test/utils/FetchMockHelper.js
import fetchMock from 'fetch-mock';

class FetchMockHelper {
  constructor(viewConf, testName) {
    this.checkViewConf(viewConf);

    this.server = require('karma-server-side'); // eslint-disable-line

    fetchMock.config.fallbackToNetwork = false;
    fetchMock.config.warnOnFallback = false;

    this.testName = testName;

    this.mockedData = {};
    this.writeToFile = false;
  }

  async getMockedData() {
    const mockedResponses = await this.server.run(this.testName, function (
      testName,
    ) {
      try {
        const fs = serverRequire('fs-extra'); // eslint-disable-line
        const path = `./test/mocked-responses/${testName}.json`;

        // Read currently available mocked responses
        if (fs.pathExistsSync(path)) {
          return fs.readJsonSync(path);
        }
        return {};
      } catch (error) {
        return error;
      }
    });

    return mockedResponses;
  }

  async getOriginalFetchResponse(url, headers) {
    // This basically disables fetch-moch, so that we can call the original fetch
    fetchMock.config.fallbackToNetwork = 'always';

    const response = await fetch(url, headers);
    let data;

    if (
      headers.headers['Content-Type'] === 'text/plain' ||
      headers.headers['content-type'] === 'text/plain'
    ) {
      data = response.text();
    } else {
      data = response.json();
    }

    // Switch fetch-mock on again
    fetchMock.config.fallbackToNetwork = false;

    return data;
  }

  async activateFetchMock() {
    this.mockedData = await this.getMockedData();

    // Since we are not using the actual mocking functionality of fetch-mock,
    // catch will intercept every call of the global fetch method
    fetchMock.catch(async (url, headers) => {
      const [requestIds, isTileData] = this.getRequestIds(url);
      let data = {};

      // Check if all the requested data is already mocked
      let isAllDataMocked = true;
      requestIds.forEach((id) => {
        if (this.mockedData[id] === undefined) {
          isAllDataMocked = false;
        }
      });

      if (isAllDataMocked) {
        requestIds.forEach((id) => {
          if (isTileData) {
            data[id] = this.mockedData[id];
          } else {
            data = this.mockedData[id];
          }
        });
      } else {
        this.writeToFile = true;
        // If there is no mocked data, load from server (specified in viewConf)
        console.warn(
          `Not all requests have been mocked. Loading ${url} from server.`,
        );
        data = await this.getOriginalFetchResponse(url, headers);
        this.addToMockedData(data, isTileData ? null : url, requestIds);
      }
      return data;
    });
  }

  addToMockedData(response, customId, requestIds) {
    if (customId === null) {
      // console.warn(ids);
      for (const id of requestIds) {
        // const id = rid.split('/')[1];
        this.mockedData[id] = response[id] === undefined ? {} : response[id];
      }
    } else {
      this.mockedData[customId] = response;
    }
  }

  async storeMockedDataToFile() {
    if (!this.writeToFile) {
      return;
    }

    const mockedResponsesJSON = JSON.stringify(this.mockedData, null, 1);

    const response = await this.server.run(
      this.testName,
      mockedResponsesJSON,
      function (testName, data) {
        try {
          // If the test is run by Travis, don't write the file
          if (!process.env.TRAVIS) {
            const fs = serverRequire('fs-extra'); // eslint-disable-line
            const path = `./test/mocked-responses/${testName}.json`;
            fs.writeFileSync(path, data);
          }
        } catch (error) {
          return error;
        }
        return null;
      },
    );

    if (response !== null) {
      console.error(
        'Could not store mocked responses',
        JSON.stringify(response),
      );
    }
  }

  async storeDataAndResetFetchMock() {
    await this.storeMockedDataToFile();
    fetchMock.reset();
  }

  getRequestIds(url) {
    const urlParts = url.split('?');

    const isTileData =
      url.includes('/tileset_info/') || url.includes('/tiles/');
    const tileIds = [];

    if (isTileData) {
      const params = new URLSearchParams(urlParts[1]);

      for (const p of params) {
        if (p[0] === 'd') {
          tileIds.push(p[1]);
        }
      }
    } else {
      tileIds.push(url);
    }

    return [tileIds, isTileData];
  }

  checkViewConf(viewConf) {
    if (JSON.stringify(viewConf).includes('"//')) {
      console.warn(
        'Please use full URLs in your view config. // is not supported and might lead to errors.',
      );
    }
  }
}

export default FetchMockHelper;


src/scripts/ClinvarTrack.js
import slugid from 'slugid';
import { isValidElement } from 'react';

const ClinvarTrack = (HGC, ...args) => {
  if (!new.target) {
    throw new Error(
      'Uncaught TypeError: Class constructor cannot be invoked without "new"',
    );
  }

  // Services
  const { tileProxy } = HGC.services;

  // Utils
  const { colorToHex, trackUtils, absToChr } = HGC.utils;

  class ClinvarTrackClass extends HGC.tracks.BedLikeTrack {
    constructor(context, options) {
      super(context, options);

      const { animate } = context;

      this.animate = animate;

      this.options = options;
      this.initOptions();

      this.offsetTop = 10;

      this.tilesToDraw = [];

      this.initSignificanceLevels();
    }

    initTile(tile) {
      tile.lollipopsGraphics = new HGC.libraries.PIXI.Graphics();
      tile.bgGraphics = new HGC.libraries.PIXI.Graphics();
      //tile.rectMaskGraphics = new HGC.libraries.PIXI.Graphics();

      tile.graphics.addChild(tile.bgGraphics);
      tile.graphics.addChild(tile.lollipopsGraphics);
      //tile.graphics.addChild(tile.rectMaskGraphics);

      //tile.rectGraphics.mask = tile.rectMaskGraphics;
      tile.clinVarData = this.interpretFieldsData(tile.tileData);
      tile.rectsForMouseOver = [];

      const bounds = this.getBoundsOfTile(tile);
      tile.tileMinX = bounds[0];
      tile.tileMaxX = bounds[1];

      tile.initialized = true;
    }

    interpretFieldsData(td) {
      const clinVarData = [];

      td.forEach((ts) => {
        const significance = ts.fields[7].replace('/', '_').toLowerCase();
        let significancesToBeDrawn = [significance];
        let numReports = [1];

        if (significance === 'conflicting_interpretations_of_pathogenicity') {
          significancesToBeDrawn = [];
          numReports = [];
          ts.fields[8].split(',').forEach((sig) => {
            const pos1 = sig.indexOf('(');
            const pos2 = sig.indexOf(')');
            significancesToBeDrawn.push(
              sig.substr(0, pos1).replace('/', '_').toLowerCase(),
            );
            numReports.push(+sig.substr(pos1 + 1, pos2 - pos1 - 1));
          });
        }

        for (var i = 0; i < significancesToBeDrawn.length; i++) {
          const entry = {
            chrom: ts.fields[0],
            posRel: +ts.fields[1],
            posAbs: ts.xStart,
            ref: ts.fields[3],
            alt: ts.fields[4],
            importance: ts.fields[5],
            goldStars: +ts.fields[6],
            significance: significance, // represents keys in this.significance
            significanceDisplay: ts.fields[7].replace(/_/g, ' '),
            significanceToBeDrawn: significancesToBeDrawn[i],
            significanceConf: ts.fields[8],
            variantType: ts.fields[9],
            origin: +ts.fields[10],
            molecularConsequence: ts.fields[11],
            diseaseName: ts.fields[12],
            hgvs: ts.fields[13],
            numReports: numReports[i], // This is the number in significanceConfs, it is 1 for not conflicting_interpretations_of_pathogenicity
            drawOrder: this.significanceLevels[significancesToBeDrawn[i]]
              .drawOrder,
            horizontalShift: 0, // will be used in case there are two reports for the same variant
          };

          clinVarData.push(entry);
        }
      });

      const clinVarDataSorted = clinVarData.sort((a, b) =>
        a.posAbs > b.posAbs ? 1 : -1,
      );
      const entriesThatPotentiallyRequireShift = {};

      clinVarDataSorted.forEach((data, index) => {
        if (index === 0) return;

        if (data.posAbs === clinVarDataSorted[index - 1].posAbs) {
          if (!entriesThatPotentiallyRequireShift[data.posAbs]) {
            entriesThatPotentiallyRequireShift[data.posAbs] = [];
            entriesThatPotentiallyRequireShift[data.posAbs].push({
              index: index - 1,
              significanceToBeDrawn:
                clinVarDataSorted[index - 1].significanceToBeDrawn,
            });
          }

          entriesThatPotentiallyRequireShift[data.posAbs].push({
            index: index,
            significanceToBeDrawn: data.significanceToBeDrawn,
          });
        }
      });

      Object.keys(entriesThatPotentiallyRequireShift).forEach((posAbs) => {
        // Group each array of object by significanceToBeDrawn
        // if they have the same posAbs but not the same significance, no shift is needed
        const grouped = entriesThatPotentiallyRequireShift[posAbs].reduce(
          function (h, obj) {
            h[obj.significanceToBeDrawn] = (
              h[obj.significanceToBeDrawn] || []
            ).concat(obj);
            return h;
          },
          {},
        );

        Object.keys(grouped).forEach((significanceToBeDrawn) => {
          const group = grouped[significanceToBeDrawn];
          if (group.length <= 1) return; // no shift required

          group.forEach((itemToShift, groupInd) => {
            clinVarDataSorted[itemToShift.index].horizontalShift = Math.trunc(
              groupInd - group.length / 2,
            );
          });
        });
      });

      clinVarData.sort((a, b) => (a.drawOrder > b.drawOrder ? 1 : -1));

      return clinVarData;
    }

    getLabelTextSprite(str) {
      const text = new HGC.libraries.PIXI.Text(str, this.labelTextOptions);
      text.interactive = true;
      text.anchor.x = 0;
      text.anchor.y = 0;
      text.visible = true;

      const pixiSprite = new HGC.libraries.PIXI.Sprite(text.texture);
      pixiSprite.width = text.getBounds().width / 2;
      pixiSprite.height = text.getBounds().height / 2;

      return pixiSprite;
    }

    initSignificanceLevels() {
      const pathogenic = {
        offsetY: 0 * this.options.levelDistance + this.offsetTop,
        label: 'Pathogenic',
        labelVisible: true,
        color: this.options.significanceColors['pathogenic'],
        colorHex: colorToHex(this.options.significanceColors['pathogenic']),
        sprite: this.getLabelTextSprite('Pathogenic'),
        drawOrder: 1,
      };

      const pathogenic_likely_pathogenic = {
        offsetY: 0.5 * this.options.levelDistance + this.offsetTop,
        label: 'Pathogenic / Likely pathogenic',
        labelVisible: false,
        color: this.options.significanceColors['pathogenic_likely_pathogenic'],
        colorHex: colorToHex(
          this.options.significanceColors['pathogenic_likely_pathogenic'],
        ),
        sprite: this.getLabelTextSprite('Pathogenic / Likely pathogenic'),
        drawOrder: 3,
      };

      const likely_pathogenic = {
        offsetY: 1 * this.options.levelDistance + this.offsetTop,
        label: 'Likely pathogenic',
        labelVisible: true,
        color: this.options.significanceColors['likely_pathogenic'],
        colorHex: colorToHex(
          this.options.significanceColors['likely_pathogenic'],
        ),
        sprite: this.getLabelTextSprite('Likely pathogenic'),
        drawOrder: 5,
      };

      const uncertain_significance = {
        offsetY: 2 * this.options.levelDistance + this.offsetTop,
        label: 'Uncertain significance',
        labelVisible: true,
        color: this.options.significanceColors['uncertain_significance'],
        colorHex: colorToHex(
          this.options.significanceColors['uncertain_significance'],
        ),
        sprite: this.getLabelTextSprite('Uncertain significance'),
        drawOrder: 7,
      };

      const likely_benign = {
        offsetY: 3 * this.options.levelDistance + this.offsetTop,
        label: 'Likely benign',
        labelVisible: true,
        color: this.options.significanceColors['likely_benign'],
        colorHex: colorToHex(this.options.significanceColors['likely_benign']),
        sprite: this.getLabelTextSprite('Likely benign'),
        drawOrder: 6,
      };

      const benign_likely_benign = {
        offsetY: 3.5 * this.options.levelDistance + this.offsetTop,
        label: 'Benign / Likely benign',
        labelVisible: false,
        color: this.options.significanceColors['benign_likely_benign'],
        colorHex: colorToHex(
          this.options.significanceColors['benign_likely_benign'],
        ),
        sprite: this.getLabelTextSprite('Benign / Likely benign'),
        drawOrder: 4,
      };

      const benign = {
        offsetY: 4 * this.options.levelDistance + this.offsetTop,
        label: 'Benign',
        labelVisible: true,
        color: this.options.significanceColors['benign'],
        colorHex: colorToHex(this.options.significanceColors['benign']),
        sprite: this.getLabelTextSprite('Benign'),
        drawOrder: 2,
      };

      const risk_factor = {
        offsetY: 4 * this.options.levelDistance + this.offsetTop,
        label: 'Risk factor',
        labelVisible: false,
        color: this.options.significanceColors['risk_factor'],
        colorHex: colorToHex(this.options.significanceColors['risk_factor']),
        sprite: this.getLabelTextSprite('Risk factor'),
        drawOrder: 0,
      };

      const conflicting_interpretations_of_pathogenicity = {
        offsetY: null,
        label: 'Conflicting interpretations of pathogenicity',
        labelVisible: false,
        color: '#0000ff',
        colorHex: colorToHex('#0000ff'),
        sprite: this.getLabelTextSprite(
          'Conflicting interpretations of pathogenicity',
        ),
        drawOrder: -1,
      };

      this.significanceLevels = {
        pathogenic,
        pathogenic_likely_pathogenic,
        likely_pathogenic,
        uncertain_significance,
        likely_benign,
        benign_likely_benign,
        benign,
        risk_factor,
        conflicting_interpretations_of_pathogenicity,
      };
    }

    initOptions() {
      this.fontSize = +this.options.fontSize;

      this.colors = {};
      this.colors['labelTextColor'] = colorToHex(this.options.labelTextColor);
      this.colors['black'] = colorToHex('#000000');
      this.colors['white'] = colorToHex('#ffffff');
      this.colors['lightgrey'] = colorToHex('#ededed');

      this.labelTextOptions = {
        fontSize: `${this.fontSize * 2}px`,
        fontFamily: this.options.fontFamily,
        fill: this.colors['labelTextColor'],
      };
    }

    /*
     * Redraw the track because the options
     * changed
     */
    rerender(options, force) {
      const strOptions = JSON.stringify(options);
      if (!force && strOptions === this.prevOptions) return;

      this.options = options;
      this.initOptions();
      this.initSignificanceLevels();

      this.prevOptions = strOptions;

      this.drawLabels();
      this.draw();

    }

    drawTile(tile) {}

    renderTile(tile) {
      // We are storing the currently visible tiles here. These will be drawn,
      // even if new tiles haven't finished loading => smoother scrolling
      this.tilesToDraw = [];
      this.visibleAndFetchedTiles().forEach((tile) => {
        this.tilesToDraw.push(tile);
      });
    }

    draw() {
      this.zoomLevel = this.calculateZoomLevel();

      this.tilesToDraw.forEach((tile) => {
        this.drawTileBackground(tile);
        this.drawLollipops(tile);
      });

    }

    drawLollipops(tile) {
      tile.lollipopsGraphics.removeChildren();
      tile.lollipopsGraphics.clear();
      tile.lollipopsGraphics.alpha = this.zoomLevel == this.maxZoom ? 1.0 : 0.5;
      tile.rectsForMouseOver = [];
      tile.svgData = [];

      const slus = this.significanceLevels['uncertain_significance'];
      const yZero = slus.offsetY + slus.sprite.height / 2;
      const circleRadius = 4;
      const circleDiameter = 2 * circleRadius;

      tile.clinVarData.forEach((cvEntry) => {
        const sl = this.significanceLevels[cvEntry.significanceToBeDrawn];
        tile.lollipopsGraphics.beginFill(sl.colorHex);
        const posX = this._xScale(cvEntry.posAbs + 0.5); // We want to draw in the middle of a nucleatide
        const posXshifted = posX + cvEntry.horizontalShift * circleDiameter;
        const posY = sl.offsetY + sl.sprite.height / 2 + 0.5;

        tile.lollipopsGraphics.drawCircle(posXshifted, posY, circleRadius);

        tile.lollipopsGraphics.drawRect(posX - 0.5, yZero, 1, posY - yZero);

        tile.rectsForMouseOver.push({
          data: cvEntry,
          rect: [
            posXshifted - circleRadius,
            posXshifted + circleRadius,
            posY - circleRadius,
            posY + circleRadius,
          ],
        });

        tile.svgData.push({
          posX: posXshifted,
          posY: posY,
          yZero: yZero,
          color: sl.color,
        });
      });
    }

    // This should be drawn only once (and not based on tiles but the whole track)
    drawTileBackground(tile) {
      tile.bgGraphics.removeChildren();
      tile.bgGraphics.clear();

      const minX = this._xScale(tile.tileMinX);
      const maxX = this._xScale(tile.tileMaxX);

      tile.bgGraphics.beginFill(this.colors['lightgrey']);

      Object.keys(this.significanceLevels).forEach((key) => {
        const sl = this.significanceLevels[key];
        if (!sl.labelVisible) return;

        tile.bgGraphics.drawRect(
          minX,
          sl.offsetY + sl.sprite.height / 2,
          maxX - minX, //this.dimensions[0],
          1,
        );
      });
    }

    drawLabels() {
      this.pForeground.clear();
      this.pForeground.removeChildren();
      let maxLabelWidth = 0;

      Object.keys(this.significanceLevels).forEach((key) => {
        const sl = this.significanceLevels[key];
        if (!sl.labelVisible) return;

        maxLabelWidth = Math.max(maxLabelWidth, sl.sprite.width);
      });

      this.pForeground.beginFill(this.colors['white']);
      this.pForeground.drawRect(
        0,
        this.offsetTop,
        maxLabelWidth + 2,
        5 * this.options.levelDistance,
      );

      Object.keys(this.significanceLevels).forEach((key) => {
        const sl = this.significanceLevels[key];
        if (!sl.labelVisible) return;
        const sprite = sl.sprite;
        sprite.position.x = 0;
        sprite.position.y = sl.offsetY;

        this.pForeground.addChild(sprite);

        // this.pForeground.drawRect(
        //   0,
        //   (i + 1) * (this.rowHeight + this.rowSpacing),
        //   this.dimensions[0],
        //   1
        // );
      });
    }

    zoomedY(yPos, kMultiplier) {}

    movedY(dY) {}

    getBoundsOfTile(tile) {
      // get the bounds of the tile
      const tileId = +tile.tileId.split('.')[1];
      const zoomLevel = +tile.tileId.split('.')[0]; //track.zoomLevel does not always seem to be up to date
      const tileWidth = +this.tilesetInfo.max_width / 2 ** zoomLevel;
      const tileMinX = this.tilesetInfo.min_pos[0] + tileId * tileWidth; // abs coordinates
      const tileMaxX = this.tilesetInfo.min_pos[0] + (tileId + 1) * tileWidth;

      return [tileMinX, tileMaxX];
    }

    /** cleanup */
    destroyTile(tile) {
      tile.bgGraphics.destroy();
      tile.lollipopsGraphics.destroy();
      tile.graphics.destroy();
      tile = null;
    }

    calculateZoomLevel() {
      // offset by 2 because 1D tiles are more dense than 2D tiles
      // 1024 points per tile vs 256 for 2D tiles
      if (!this.tilesetInfo) return;

      const xZoomLevel = tileProxy.calculateZoomLevel(
        this._xScale,
        this.tilesetInfo.min_pos[0],
        this.tilesetInfo.max_pos[0],
      );

      let zoomLevel = Math.min(xZoomLevel, this.maxZoom);
      zoomLevel = Math.max(zoomLevel, 0);

      return zoomLevel;
    }

    isPointInRectangle(rect, point) {
      if (
        rect[0] < point[0] &&
        rect[1] > point[0] &&
        rect[2] < point[1] &&
        rect[3] > point[1]
      ) {
        return true;
      }
      return false;
    }

    getMouseOverHtml(trackX, trackY) {
      if (!this.tilesetInfo) {
        return '';
      }

      const point = [trackX, trackY];

      for (const tile of this.visibleAndFetchedTiles()) {
        for (let i = 0; i < tile.rectsForMouseOver.length; i++) {
          const rect = tile.rectsForMouseOver[i].rect;
          const data = tile.rectsForMouseOver[i].data;
          if (this.isPointInRectangle(rect, point)) {
            let goldStarString = '';
            for (var i = 0; i < data.goldStars; i++) {
              goldStarString += "<span style='color:#ebdc00'>&#9733;</span>";
            }
            for (var i = 0; i < 4 - data.goldStars; i++) {
              goldStarString += "<span style='color:#bdbdbd'>&#9734;</span>";
            }
            const posRel = data.posRel.toLocaleString('en-US', {
              minimumFractionDigits: 0,
            });
            return `
                <div style='text-align: center'>
                  <div><b>${data.chrom}:${posRel}</b></div>
                  <div>${data.ref} &#8658; ${data.alt}</div>
                  <div>Clinvar review status: ${goldStarString}</div>
                  <div>Clinical significance: <b>${data.significanceDisplay}</b></div>
                </div>
              `;
          }
        }
      }
    }

    exportSVG() {
      let track = null;
      let base = null;

      base = document.createElement('g');
      track = base;

      const clipPathId = slugid.nice();

      const gClipPath = document.createElement('g');
      gClipPath.setAttribute('style', `clip-path:url(#${clipPathId});`);

      track.appendChild(gClipPath);

      // define the clipping area as a polygon defined by the track's
      // dimensions on the canvas
      const clipPath = document.createElementNS(
        'http://www.w3.org/2000/svg',
        'clipPath',
      );
      clipPath.setAttribute('id', clipPathId);
      track.appendChild(clipPath);

      const clipPolygon = document.createElementNS(
        'http://www.w3.org/2000/svg',
        'polygon',
      );
      clipPath.appendChild(clipPolygon);

      clipPolygon.setAttribute(
        'points',
        `${this.position[0]},${this.position[1]} ` +
          `${this.position[0] + this.dimensions[0]},${this.position[1]} ` +
          `${this.position[0] + this.dimensions[0]},${
            this.position[1] + this.dimensions[1]
          } ` +
          `${this.position[0]},${this.position[1] + this.dimensions[1]} `,
      );

      const output = document.createElement('g');

      output.setAttribute(
        'transform',
        `translate(${this.position[0]},${this.position[1]})`,
      );

      gClipPath.appendChild(output);

      // Horizontal lines
      Object.keys(this.significanceLevels).forEach((sl) => {
        if (this.significanceLevels[sl].labelVisible) {
          const g = document.createElement('g');
          const t = document.createElement('line');
          t.setAttribute('x1', 0);
          t.setAttribute('x2', this.dimensions[0]);
          t.setAttribute('y1', 0);
          t.setAttribute('y2', 0);
          t.setAttribute('stroke', this.options.labelTextColor);
          t.setAttribute('stroke-width', 0.5);

          g.setAttribute('transform', `scale(1,1)`);

          const spHeight = this.significanceLevels[sl].sprite.height;

          g.appendChild(t);
          g.setAttribute(
            'transform',
            `translate(0,${
              this.significanceLevels[sl].offsetY + spHeight / 2
            })scale(1,1)`,
          );
          output.appendChild(g);
        }
      });

      // Labels
      const totalHeight = 5 * this.options.levelDistance + 5;

      let maxLabelWidth = 0;

      Object.keys(this.significanceLevels).forEach((sl) => {
        if (this.significanceLevels[sl].labelVisible) {
          maxLabelWidth = Math.max(
            maxLabelWidth,
            this.significanceLevels[sl].sprite.width,
          );
        }
      });

      // Lollipops
      this.visibleAndFetchedTiles().forEach((tile) => {
        tile.svgData.forEach((entry) => {
          const g = document.createElement('g');
          const t = document.createElement('circle');
          t.setAttribute('cx', entry.posX);
          t.setAttribute('cy', entry.posY);
          t.setAttribute('r', 4);

          g.setAttribute('transform', `scale(1,1)`);

          t.setAttribute('fill', entry.color);
          g.appendChild(t);

          const l = document.createElement('line');
          l.setAttribute('x1', entry.posX);
          l.setAttribute('x2', entry.posX);
          l.setAttribute('y1', entry.posY);
          l.setAttribute('y2', entry.yZero);
          l.setAttribute('stroke', entry.color);
          l.setAttribute('stroke-width', 1);
          g.appendChild(l);

          output.appendChild(g);
        });
      });

      const gLabelBackground = document.createElement('g');
      const rLabelBackground = document.createElement('path');
      const dLabelBackground = `M 0 0 H ${
        maxLabelWidth + 3
      } V ${totalHeight} H 0 Z`;
      rLabelBackground.setAttribute('d', dLabelBackground);
      rLabelBackground.setAttribute('fill', 'white');
      rLabelBackground.setAttribute('opacity', '1');
      gLabelBackground.appendChild(rLabelBackground);
      output.appendChild(gLabelBackground);

      Object.keys(this.significanceLevels).forEach((sl) => {
        if (this.significanceLevels[sl].labelVisible) {
          const g = document.createElement('g');
          const t = document.createElement('text');
          t.setAttribute('text-anchor', 'start');
          t.setAttribute('font-family', this.options.fontFamily);
          t.setAttribute('font-size', `${this.fontSize}px`);
          //t.setAttribute("font-weight", "bold");

          g.setAttribute('transform', `scale(1,1)`);

          t.setAttribute('fill', this.options.labelTextColor);
          t.innerHTML = this.significanceLevels[sl].label;

          g.appendChild(t);
          g.setAttribute(
            'transform',
            `translate(0,${
              this.significanceLevels[sl].offsetY + this.offsetTop
            })scale(1,1)`,
          );
          output.appendChild(g);
        }
      });

      return [base, base];
    }
  }
  return new ClinvarTrackClass(...args);
};

const icon =
  '<svg width="20" height="20" xmlns="http://www.w3.org/2000/svg"><path fill="#fff" d="M-1-1h22v22H-1z"/><g><path stroke="#007fff" stroke-width="1.5" fill="#007fff" d="M-.667-.091h5v20.167h-5z"/><path stroke-width="1.5" stroke="#e8e500" fill="#e8e500" d="M5.667.242h5v20.167h-5z"/><path stroke-width="1.5" stroke="#ff0038" fill="#ff0038" d="M15.833.076h5v20.167h-5z"/><path stroke="green" stroke-width="1.5" fill="green" d="M10.833-.258H14.5v20.167h-3.667z"/></g></svg>';

// default
ClinvarTrack.config = {
  type: 'horizontal-clinvar',
  datatype: ['bedlike'],
  local: false,
  orientation: '1d-horizontal',
  thumbnail: new DOMParser().parseFromString(icon, 'text/xml').documentElement,
  availableOptions: [
    'fontSize',
    'fontFamily',
    'labelTextColor',
    'levelDistance',
    'significanceColors',
  ],
  defaultOptions: {
    fontSize: 10,
    fontFamily: 'Arial',
    labelTextColor: '#888888',
    levelDistance: 20,
    significanceColors: {
      pathogenic: '#D55E00',
      pathogenic_likely_pathogenic: '#CC79A7',
      likely_pathogenic: '#E69F00',
      uncertain_significance: '#808080',
      likely_benign: '#56B4E9',
      benign_likely_benign: '#0072B2',
      benign: '#009E73',
      risk_factor: '#999999',
    },
  },
};

export default ClinvarTrack;
